From b8d81c5402eab8cab9ff98d3589e1d87719c2368 Mon Sep 17 00:00:00 2001
From: Pete Beardmore <pete.beardmore@msn.com>
Date: Sat, 24 Mar 2018 23:59:59 +0000
Subject: embeddedartwidgetbar, add 'autoselect all selected' option

-optionally auto select collapsed containers when all contents
(album/songs) are selected
-perform selection on collapse toggle only
---
 quodlibet/quodlibet/ext/covers/embeddedart_widgetbar.py | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/quodlibet/quodlibet/ext/covers/embeddedart_widgetbar.py b/quodlibet/quodlibet/ext/covers/embeddedart_widgetbar.py
index 8cb4a8a2c..f6658549e 100644
--- a/quodlibet/quodlibet/ext/covers/embeddedart_widgetbar.py
+++ b/quodlibet/quodlibet/ext/covers/embeddedart_widgetbar.py
@@ -41,6 +41,8 @@ class Config(object):
     size_in_label = BoolConfProp(_config, "size_in_label", False)
     size_in_tooltip = BoolConfProp(_config, "size_in_tooltip", True)
     collapsed_view = BoolConfProp(_config, "collapsed_view", False)
+    autoselect_all_selected =\
+         BoolConfProp(_config, "autoselect_all_selected", False)
 
 
 CONFIG = Config()
@@ -193,6 +195,8 @@ class ImageWidget(Gtk.HBox):
             return self.is_selected
 
     def collapsed_select_check(self, album=None):
+        autoselect = CONFIG.autoselect_all_selected
+
         for k, g in groupby(self.nested,
                             lambda iw: iw.song['album']):
             if album and not k == album:
@@ -202,12 +206,18 @@ class ImageWidget(Gtk.HBox):
             selected_song_count = len([iw for iw in group
                                           if self.nested_active[iw]])
 
+            all_selected = selected_song_count == song_count
             if (selected_song_count >= song_count - 1):
                 # check album selection state
                 album_cb = self.__get_nested_album(self, k)
                 self.update_lock = True
-                album_cb.set_active(selected_song_count == song_count)
+                album_cb.set_active(all_selected)
                 self.update_lock = False
+            if autoselect:
+                autoselect = all_selected
+
+        if not album and autoselect:
+            self.highlight_toggle(True)
 
     def collapsed(self, visible):
         if visible:
@@ -1208,6 +1218,9 @@ class EmbeddedArtWidgetBarPlugin(UserInterfacePlugin, EventPlugin):
 
         # toggles
         toggles = [
+            (plugin_id + '_autoselect_all_selected',
+             _("Autoselect collapsed containers with all selected"),
+             None, False, lambda *_: None),
             (plugin_id + '_name_in_label', _("Show image name in label"),
              None, False, lambda *x: self.__embeddedart_box.update_labels()),
             (plugin_id + '_size_in_label', _("Show image size in label"),
-- 
2.14.1

